lib/record_store/zone.rb

```diff
@@ -272,6 +273,26 @@ def validate_provider_can_handle_zone_records
       end
     end
 
+    def validate_zone_record_not_shadowed
+      nameserver_records = records
+        .select { |record| record.is_a?(Record::NS) && name != record.fqdn }
+        .map { |record| record.fqdn.delete_suffix(".") }
+        .uniq
+
+      nameserver_records.each do |ns_record|
+        selected_records = records.reject do |record|
+          record.is_a?(Record::NS) && \
+            record.fqdn.delete_suffix(".") == ns_record
+        end
+        selected_records.each do |record|
```

Comment:
total nitpick, lean into collection pipelines and inline this temp variable:

```suggestion
       records.reject do |record|
          record.is_a?(Record::NS) && \
            record.fqdn.delete_suffix(".") == ns_record
       end.each do |record|
```

Reply:
I really love that but I'm going to skip it for readability